### PR TITLE
[5.3] AST: isEquivalentToExtendedContext() should consider use patterns of @_originallyDefinedIn

### PR DESCRIPTION
**Pull request by nkcsgexi**
When a nominal type and its extensions coexist in the same module, mangling may
use the nominal type as the context for extension members. We should preserve this
mechanism for types marked as @_originallyDefinedIn to maintain ABI stability.

rdar://64538537

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please test

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -1325,7 +1325,12 @@ bool ExtensionDecl::isConstrainedExtension() const {
 
 bool ExtensionDecl::isEquivalentToExtendedContext() const {
   auto decl = getExtendedNominal();
-  return getParentModule() == decl->getParentModule()
+  auto extendDeclfromSameModule =
```
Comment (slavapestov) [inline lib/AST/Decl.cpp:1328]: Nitpick: capitalize the "from"

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -1325,7 +1325,12 @@ bool ExtensionDecl::isConstrainedExtension() const {
...
+    getParentModule() == decl->getParentModule() ||
+    // if the extended type was defined in the same module with the extension,
+    // we should consider them as the same module to preserve ABI stability.
+    decl->getAlternateModuleName() == getParentModule()->getNameStr();
```
Comment (slavapestov) [inline lib/AST/Decl.cpp:1332]: Should we even check "getParentModule() == decl->getParentModule()" if an alternate module name is defined?

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -1325,7 +1325,12 @@ bool ExtensionDecl::isConstrainedExtension() const {
...
+  return extendDeclfromSameModule
     && !isConstrainedExtension()
     && !getDeclaredInterfaceType()->isExistentialType();
```
Comment (slavapestov) [inline lib/AST/Decl.cpp:1335]: This is better written as "isa<ProtocolDecl>(decl)"

**Review by slavapestov — COMMENTED** [review]
Can you add a SILGen test for the mangling as well?

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@slavapestov Thank you for taking a look! The PR for addressing the code review comments was cherry-picked.

**Review by tkremenek — APPROVED** [review]

**Pull request merged**